File: STOD_Calculation_Pipeline/cpp_backend/sod_canary.cpp
#include <iostream>
#include <vector>
#include <iomanip>
#include <cmath>
#include <algorithm>
#include <unordered_set>
#include <string>
#include "sod_logic.hpp"

using namespace std;

// ---------------------------------------------------------
// Visual Logger (Replicates Python format_stod_log)
// ---------------------------------------------------------
void log_stod_run(const vector<Point>& path_a, const vector<Point>& path_b) {
    size_t len_a = path_a.size();
    size_t len_b = path_b.size();
    size_t n = (len_a < len_b) ? len_a : len_b;

    cout << "\n--- STOD Detailed Uncanceled Contribution Log ---" << endl;
    cout << "Level | Side A (row, col)    | Side B (row, col)    | Canc | Unc  | Unc * level" << endl;
    cout << "---------------------------------------------------------------------------------------" << endl;

    if (n == 0) {
        cout << "\nTotal Stability Score: 0\nStatus: Empty paths\n";
        return;
    }

    // 1) Simulation phase – identical to calculate_stod_pair
    unordered_set<int> a_rows, a_cols;
    unordered_set<int> b_rows, b_cols;
    a_rows.reserve(n); a_cols.reserve(n);
    b_rows.reserve(n); b_cols.reserve(n);

    bool terminated = false;
    size_t final_level = 0;

    for (size_t level = 0; level < n; ++level) {
        final_level = level;
        const Point& pA = path_a[level];
        const Point& pB = path_b[level];

        a_rows.insert(pA.r); a_cols.insert(pA.c);
        b_rows.insert(pB.r); b_cols.insert(pB.c);

        // Termination Condition: Check if ANY level (0 to current) on either side
        // now has BOTH coordinates covered by the other side's accumulated visits.
        for (size_t check_level = 0; check_level <= level; ++check_level) {
            const Point& checkA = path_a[check_level];
            const Point& checkB = path_b[check_level];
            
            bool term_A = (b_rows.count(checkA.r) && b_cols.count(checkA.c));
            bool term_B = (a_rows.count(checkB.r) && a_cols.count(checkB.c));
            
            if (term_A || term_B) {
                terminated = true;
                break;
            }
        }
        
        if (terminated) {
            break;
        }
    }

    // 2) Scoring + logging phase – use FINAL sets (global cancellations)
    double total_score = 0.0;

    for (size_t level = 0; level <= final_level; ++level) {
        const Point& pA = path_a[level];
        const Point& pB = path_b[level];

        bool canc_Ar = (b_rows.count(pA.r) > 0);
        bool canc_Ac = (b_cols.count(pA.c) > 0);
        bool canc_Br = (a_rows.count(pB.r) > 0);
        bool canc_Bc = (a_cols.count(pB.c) > 0);

        int canc = (int)canc_Ar + (int)canc_Ac + (int)canc_Br + (int)canc_Bc;
        int unc  = 4 - canc;
        double lvl_score = static_cast<double>(level) * static_cast<double>(unc);
        total_score += lvl_score;

        // pretty printing with parentheses around canceled components
        cout << setw(5) << level << " | "
             << "(" << (canc_Ar ? "(" : " ") << pA.r << (canc_Ar ? ")" : " ") << ", "
             << (canc_Ac ? "(" : " ") << pA.c << (canc_Ac ? ")" : " ") << ")   | "
             << "(" << (canc_Br ? "(" : " ") << pB.r << (canc_Br ? ")" : " ") << ", "
             << (canc_Bc ? "(" : " ") << pB.c << (canc_Bc ? ")" : " ") << ")   | "
             << "  " << canc << "  |  " << unc << "   | "
             << unc << " * " << level << " = " << lvl_score << endl;
    }

    cout << "\nTotal Stability Score: " << total_score << endl;
    if (terminated) {
        cout << "Status: Terminated (Type 0)" << endl;
    } else {
        cout << "Status: Unterminated" << endl;
    }
}


// Generate Serpentine Path (6x6 grid, matching Python version)
vector<Point> generate_serpentine_path(int grid_size) {
    vector<Point> path;
    for (int r = 0; r < grid_size; ++r) {
        if (r % 2 == 0) {
            for (int c = 0; c < grid_size; ++c) {
                path.push_back({r, c});
            }
        } else {
            for (int c = grid_size - 1; c >= 0; --c) {
                path.push_back({r, c});
            }
        }
    }
    return path;
}

// Cyclic slice: take 'length' points along the serpentine cycle starting at start_idx
vector<Point> cyclic_slice(const vector<Point>& full_path, int start_idx, int length) {
    int N = full_path.size();
    vector<Point> result;
    for (int i = 0; i < length; ++i) {
        int idx = (start_idx + i) % N;
        result.push_back(full_path[idx]);
    }
    return result;
}

int main() {
    cout << "============================================================" << endl;
    cout << "   C++ STOD/FINSTOD CANARY TEST (SERPENTINE)" << endl;
    cout << "============================================================" << endl;

    // Setup matching Python version: 6x6 grid, indices 5 and 10, LENGTH=11
    int grid_size = 6;
    int TIME_UNITS = 10;
    int LENGTH = TIME_UNITS + 1;  // 11 points
    
    vector<Point> full_snake = generate_serpentine_path(grid_size);
    
    int idx_A = 5;
    int idx_B = 10;
    
    // Finite "time-window" trajectories with wrap-around (periodic snake)
    vector<Point> traj_A = cyclic_slice(full_snake, idx_A, LENGTH);
    vector<Point> traj_B = cyclic_slice(full_snake, idx_B, LENGTH);
    
    // Reversed trajectories (final -> initial): Level 0 becomes the final cell
    vector<Point> traj_A_rev = traj_A;
    reverse(traj_A_rev.begin(), traj_A_rev.end());
    vector<Point> traj_B_rev = traj_B;
    reverse(traj_B_rev.begin(), traj_B_rev.end());

    // Reversibility check
    vector<Point> traj_A_check = traj_A_rev;
    reverse(traj_A_check.begin(), traj_A_check.end());
    vector<Point> traj_B_check = traj_B_rev;
    reverse(traj_B_check.begin(), traj_B_check.end());
    bool check_A = (traj_A == traj_A_check);
    bool check_B = (traj_B == traj_B_check);

    // Compute STOD (forward) and FINSTOD (reversed)
    SmodResult res_stod = calculate_stod_pair(traj_A, traj_B);
    SmodResult res_finstod = calculate_stod_pair(traj_A_rev, traj_B_rev);

    // Expected values from Python canary
    // Note: FINSTOD terminates at level 3 (not level 5) because A's level 0 point (2,3)
    // becomes fully covered when B visits row 2 at level 3 (B already had col 3 from level 0)
    double expected_stod_score = 21.0;
    double expected_finstod_score = 15.0;  // Corrected: terminates at level 3, score = 0+3+6+6 = 15
    int expected_stod_type = 0;  // T (Terminated)
    int expected_finstod_type = 0;   // T (Terminated)

    cout << "\n--- FULL SERPENTINE PATH (Size " << grid_size << "x" << grid_size << ", " 
         << full_snake.size() << " cells) ---" << endl;
    cout << "Complete path (row, col) sequence:" << endl;
    for (size_t i = 0; i < full_snake.size(); ++i) {
        cout << "  Index " << setw(2) << i << ": (" << full_snake[i].r << ", " << full_snake[i].c << ")" << endl;
    }
    cout << "\nSelected cells for testing:" << endl;
    cout << "  Cell A: Index " << idx_A << " = (" << full_snake[idx_A].r << ", " << full_snake[idx_A].c << ")" << endl;
    cout << "  Cell B: Index " << idx_B << " = (" << full_snake[idx_B].r << ", " << full_snake[idx_B].c << ")" << endl;

    cout << "\n--- REVERSIBILITY VERIFICATION ---" << endl;
    cout << "Check Reverse Identity (Path A): " << (check_A ? "YES" : "NO") 
         << " (traj_A == traj_A_rev[::-1])" << endl;
    cout << "Check Reverse Identity (Path B): " << (check_B ? "YES" : "NO")
         << " (traj_B == traj_B_rev[::-1])" << endl;
    if (check_A && check_B) {
        cout << "  ✓ Reversibility verified: forward and reversed paths are exact inverses" << endl;
    } else {
        cout << "  WARNING: Reversibility check failed!" << endl;
    }

    cout << "\n--- FORWARD PATHS (STOD inputs) ---" << endl;
    cout << "Path A forward (len=" << traj_A.size() << "): ";
    for (size_t i = 0; i < traj_A.size(); ++i) {
        cout << "(" << traj_A[i].r << "," << traj_A[i].c << ")";
        if (i < traj_A.size() - 1) cout << " ";
    }
    cout << endl;
    
    cout << "Path B forward (len=" << traj_B.size() << "): ";
    for (size_t i = 0; i < traj_B.size(); ++i) {
        cout << "(" << traj_B[i].r << "," << traj_B[i].c << ")";
        if (i < traj_B.size() - 1) cout << " ";
    }
    cout << endl;

    cout << "\n--- REVERSED PATHS (FINSTOD inputs; Level 0 is final cell) ---" << endl;
    cout << "Path A reversed (len=" << traj_A_rev.size() << "): ";
    for (size_t i = 0; i < traj_A_rev.size(); ++i) {
        cout << "(" << traj_A_rev[i].r << "," << traj_A_rev[i].c << ")";
        if (i < traj_A_rev.size() - 1) cout << " ";
    }
    cout << endl;
    
    cout << "Path B reversed (len=" << traj_B_rev.size() << "): ";
    for (size_t i = 0; i < traj_B_rev.size(); ++i) {
        cout << "(" << traj_B_rev[i].r << "," << traj_B_rev[i].c << ")";
        if (i < traj_B_rev.size() - 1) cout << " ";
    }
    cout << endl;

    cout << "\n" << string(80, '=') << endl;
    cout << "STOD CALCULATION (Forward Paths)" << endl;
    cout << string(80, '=') << endl;
    cout << "Using forward trajectories: traj_A vs traj_B" << endl;
    cout << "This tests the STOD logic function (calculate_stod_pair) with forward paths" << endl;
    cout << "as it will be used in the STOD calculation in the pipeline." << endl;
    cout << "STOD (Forward) A vs B - Result: Type " << res_stod.type_code << endl;
    log_stod_run(traj_A, traj_B);
    
    cout << "\n" << string(80, '=') << endl;
    cout << "FINSTOD CALCULATION (Reversed Paths)" << endl;
    cout << string(80, '=') << endl;
    cout << "Using reversed trajectories: traj_A_rev vs traj_B_rev" << endl;
    cout << "This tests the STOD logic function (calculate_stod_pair) with reversed paths" << endl;
    cout << "as it will be used in the FINSTOD calculation in the pipeline." << endl;
    cout << "FINSTOD (Reversed) A_rev vs B_rev - Result: Type " << res_finstod.type_code << endl;
    log_stod_run(traj_A_rev, traj_B_rev);

    // Validation
    bool stod_score_ok = (abs(res_stod.score - expected_stod_score) < 1e-6);
    bool stod_type_ok = (res_stod.type_code == expected_stod_type);
    bool finstod_score_ok = (abs(res_finstod.score - expected_finstod_score) < 1e-6);
    bool finstod_type_ok = (res_finstod.type_code == expected_finstod_type);

    cout << "\n" << string(80, '=') << endl;
    cout << "VALIDATION RESULTS" << endl;
    cout << string(80, '=') << endl;
    cout << "This validates that:" << endl;
    cout << "  1. The STOD logic function (calculate_stod_pair) works correctly" << endl;
    cout << "  2. Both STOD (forward) and FINSTOD (reversed) produce expected results" << endl;
    cout << "  3. These are the SAME functions used in the pipeline calculations" << endl;
    cout << string(80, '=') << endl;
    cout << "\nSTOD Validation (Forward Paths):" << endl;
    cout << "  - Score Match: " << (stod_score_ok ? "YES" : "NO") 
         << " (Expected: " << expected_stod_score << ", Got: " << res_stod.score << ")" << endl;
    cout << "  - Type Match: " << (stod_type_ok ? "YES" : "NO")
         << " (Expected: " << expected_stod_type << ", Got: " << res_stod.type_code << ")" << endl;
    if (stod_score_ok && stod_type_ok) {
        cout << "  ✓ STOD calculation verified: STOD logic works correctly with forward paths" << endl;
    }
    cout << "\nFINSTOD Validation (Reversed Paths):" << endl;
    cout << "  - Score Match: " << (finstod_score_ok ? "YES" : "NO")
         << " (Expected: " << expected_finstod_score << ", Got: " << res_finstod.score << ")" << endl;
    cout << "  - Type Match: " << (finstod_type_ok ? "YES" : "NO")
         << " (Expected: " << expected_finstod_type << ", Got: " << res_finstod.type_code << ")" << endl;
    if (finstod_score_ok && finstod_type_ok) {
        cout << "  ✓ FINSTOD calculation verified: STOD logic works correctly with reversed paths" << endl;
    }

    if (stod_score_ok && stod_type_ok && finstod_score_ok && finstod_type_ok) {
        cout << "\n[SUCCESS] C++ STOD/FINSTOD Canary Passed." << endl;
        return 0;
    } else {
        cout << "\n[FAILURE] C++ STOD/FINSTOD Canary Failed!" << endl;
        return 1;
    }
}

File: STOD_Calculation_Pipeline/pipeline_core/sod_canary.py
# -*- coding: utf-8 -*-
"""
STOD/FINSTOD Canary (Serpentine Test)
========================================
Validates the Python/Numba STOD/FINSTOD core before running the full C++ pipeline.
Uses the serpentine canary test with expected values: STOD=21, FINSTOD=15
"""
import sys
import numpy as np

from pipeline_core.sod_logic import (
    stod_pair_python,
    stod_pair_numba,
    TYPE_NAME
)

def format_stod_log(levels_data, total_score_unc, label="STOD pair"):
    """
    Log for a STOD pair (used by both STOD and FINSTOD):
      - shows which pair we are comparing (label),
      - shows paths with parentheses around canceled row/col components,
      - per level: #canceled, #uncanceled, 'Unc * level = ...',
      - at the bottom: total Σ (Unc * level).
    """
    lines = []
    lines.append(f"=== {label} ===")
    lines.append("--- STOD Detailed Uncanceled Contribution Log ---\n")
    header = (
        f"{'Level':^5} | {'Side A (r,c)':^20} | "
        f"{'Side B (r,c)':^20} | {'Canc':^4} | {'Unc':^4} | {'Unc * level':^18}"
    )
    lines.append(header)
    lines.append("-" * len(header))

    for d in levels_data:
        lvl = d["level"]
        rA, cA = d["rA"], d["cA"]
        rB, cB = d["rB"], d["cB"]

        canc_count = d["canc_count"]
        unc_count  = d["uncanc_count"]
        contrib    = d["unc_contrib"]

        # Parenthesis around canceled row / col
        sA_r = f"({rA:>2})" if d["canc_Ar"] else f" {rA:>2} "
        sA_c = f"({cA:>2})" if d["canc_Ac"] else f" {cA:>2} "
        sB_r = f"({rB:>2})" if d["canc_Br"] else f" {rB:>2} "
        sB_c = f"({cB:>2})" if d["canc_Bc"] else f" {cB:>2} "

        sideA_str = f"({sA_r}, {sA_c})"
        sideB_str = f"({sB_r}, {sB_c})"

        lines.append(
            f"{lvl:^5} | {sideA_str:^20} | {sideB_str:^20} | "
            f"{canc_count:^4d} | {unc_count:^4d} | "
            f"{unc_count} * {lvl} = {contrib:.1f}"
        )

    lines.append("")
    lines.append(f"Total (Σ Unc * level) = {total_score_unc:.1f}")
    return "\n".join(lines)

def run_sod_canary_test():
    """
    Periodic serpentine canary:
      - The serpentine path is a cycle: after the last cell, it continues at the first cell.
      - We emulate "finite time" by taking a finite number of steps (TIME_UNITS),
        with 1 step per time unit.

    STOD: compare forward finite trajectories (A vs B).
    FINSTOD: compare reversed finite trajectories (A_rev vs B_rev), so Level 0 is the final cell.
    """
    print("=" * 80)
    print("             SERPENTINE CANARY TEST (STOD/FINSTOD)")
    print("=" * 80)
    
    size = 6
    TIME_UNITS = 10          # 10 time units
    STEPS = TIME_UNITS       # 1 step per time unit
    LENGTH = STEPS + 1       # positions include the start, so steps+1 points

    def gen_serpentine(s=6):
        path = []
        for r in range(s):
            cols = range(s) if r % 2 == 0 else range(s - 1, -1, -1)
            for c in cols:
                path.append((r, c))
        return np.array(path, dtype=np.int64)

    def cyclic_slice(full_path, start_idx, length):
        """
        Take 'length' points along the serpentine cycle starting at start_idx.
        """
        N = full_path.shape[0]
        idxs = (start_idx + np.arange(length)) % N
        return full_path[idxs]

    full_snake = gen_serpentine(size)
    Ncells = full_snake.shape[0]  # = size*size

    # Display the full serpentine path
    print(f"\n--- FULL SERPENTINE PATH (Size {size}x{size}, {Ncells} cells) ---")
    print("Complete path (row, col) sequence:")
    for i, (r, c) in enumerate(full_snake):
        print(f"  Index {i:2d}: ({r}, {c})")
    print()

    # Pick two source cells
    idx_A, idx_B = 5, 10
    start_A_rc, start_B_rc = full_snake[idx_A], full_snake[idx_B]
    print(f"Selected cells for testing:")
    print(f"  Cell A: Index {idx_A} = {start_A_rc}")
    print(f"  Cell B: Index {idx_B} = {start_B_rc}")

    # Finite "time-window" trajectories with wrap-around (periodic snake)
    traj_A = cyclic_slice(full_snake, idx_A, LENGTH)
    traj_B = cyclic_slice(full_snake, idx_B, LENGTH)

    # Reversed trajectories (final -> initial): Level 0 becomes the final cell
    traj_A_rev = traj_A[::-1].copy()
    traj_B_rev = traj_B[::-1].copy()

    # Compute STOD (forward) and FINSTOD (reversed)
    ti, si, term_i, li = stod_pair_python(traj_A, traj_B)
    ts, ss, term_s, ls = stod_pair_python(traj_A_rev, traj_B_rev)

    # Strong reverse checks: the reverses should be exact inverses
    check_A = np.array_equal(traj_A, traj_A_rev[::-1])
    check_B = np.array_equal(traj_B, traj_B_rev[::-1])

    # Print detailed report
    print(f"\n--- TEST PARAMETERS ---")
    print(f"Grid size: {size}x{size} (Ncells={Ncells})")
    print(f"Time units: {TIME_UNITS} | 1 step/unit => STEPS={STEPS} | Path length={LENGTH} points")
    print(f"\n--- REVERSIBILITY VERIFICATION ---")
    print(f"Check Reverse Identity (Path A): {check_A} (traj_A == traj_A_rev[::-1])")
    print(f"Check Reverse Identity (Path B): {check_B} (traj_B == traj_B_rev[::-1])")
    if not (check_A and check_B):
        print("  WARNING: Reversibility check failed!")
    else:
        print("  ✓ Reversibility verified: forward and reversed paths are exact inverses")
    print(f"\n--- FORWARD PATHS (STOD inputs) ---")
    print(f"Path A forward (len={len(traj_A)}): {traj_A.tolist()}")
    print(f"Path B forward (len={len(traj_B)}): {traj_B.tolist()}")
    print(f"\n--- REVERSED PATHS (FINSTOD inputs; Level 0 is final cell) ---")
    print(f"Path A reversed (len={len(traj_A_rev)}): {traj_A_rev.tolist()}")
    print(f"Path B reversed (len={len(traj_B_rev)}): {traj_B_rev.tolist()}")
    print(f"\nForward ends: A_end={traj_A[-1]} | B_end={traj_B[-1]}")
    print("Note: In FINSTOD (Reversed), Level 0 corresponds to A_end and B_end.")
    
    print("\n" + "="*80)
    print("STOD CALCULATION (Forward Paths)")
    print("="*80)
    print("Using forward trajectories: traj_A vs traj_B")
    print("This tests the STOD logic function (stod_pair_python) with forward paths")
    print("as it will be used in the STOD calculation in the pipeline.")
    print(format_stod_log(li, si, label=f"STOD (Forward) A vs B - Result: {TYPE_NAME[ti]}"))
    
    print("\n" + "="*80)
    print("FINSTOD CALCULATION (Reversed Paths)")
    print("="*80)
    print("Using reversed trajectories: traj_A_rev vs traj_B_rev")
    print("This tests the STOD logic function (stod_pair_python) with reversed paths")
    print("as it will be used in the FINSTOD calculation in the pipeline.")
    print(format_stod_log(ls, ss, label=f"FINSTOD (Reversed) A_rev vs B_rev - Result: {TYPE_NAME[ts]}"))

    # Numba validation
    ti_nb, si_nb = stod_pair_numba(traj_A, traj_B)
    ts_nb, ss_nb = stod_pair_numba(traj_A_rev, traj_B_rev)

    # Expected values
    # Note: FINSTOD terminates at level 3 (not level 5) because A's level 0 point (2,3)
    # becomes fully covered when B visits row 2 at level 3 (B already had col 3 from level 0)
    expected_stod_score = 21.0
    expected_finstod_score = 15.0  # Corrected: terminates at level 3, score = 0+3+6+6 = 15

    # Validation
    stod_py_nb_match = (ti == ti_nb) and (abs(si - si_nb) < 1e-9)
    finstod_py_nb_match = (ts == ts_nb) and (abs(ss - ss_nb) < 1e-9)
    stod_expected_match = abs(si - expected_stod_score) < 1e-9
    finstod_expected_match = abs(ss - expected_finstod_score) < 1e-9

    print("\n" + "="*80)
    print("VALIDATION RESULTS")
    print("="*80)
    print("This validates that:")
    print("  1. The STOD logic function (stod_pair_python) works correctly")
    print("  2. The Numba-accelerated version (stod_pair_numba) matches Python")
    print("  3. Both STOD (forward) and FINSTOD (reversed) produce expected results")
    print("  4. These are the SAME functions used in the pipeline calculations")
    print("="*80)
    print(f"\nSTOD Validation (Forward Paths):")
    print(f"  - Python/Numba Match: {stod_py_nb_match} (Python: {si:.1f}, Numba: {si_nb:.1f})")
    print(f"  - Expected Score Match: {stod_expected_match} (Expected: {expected_stod_score:.1f}, Got: {si:.1f})")
    print(f"  - Type: {TYPE_NAME[ti]} ({ti})")
    if stod_py_nb_match and stod_expected_match:
        print(f"  ✓ STOD calculation verified: STOD logic works correctly with forward paths")
    
    print(f"\nFINSTOD Validation (Reversed Paths):")
    print(f"  - Python/Numba Match: {finstod_py_nb_match} (Python: {ss:.1f}, Numba: {ss_nb:.1f})")
    print(f"  - Expected Score Match: {finstod_expected_match} (Expected: {expected_finstod_score:.1f}, Got: {ss:.1f})")
    print(f"  - Type: {TYPE_NAME[ts]} ({ts})")
    if finstod_py_nb_match and finstod_expected_match:
        print(f"  ✓ FINSTOD calculation verified: STOD logic works correctly with reversed paths")

    if stod_py_nb_match and finstod_py_nb_match and stod_expected_match and finstod_expected_match:
        print("\n[SUCCESS] STOD/FINSTOD Canary Passed.")
        return 0
    else:
        print("\n[FAILURE] STOD/FINSTOD Canary Failed!")
        return 1

if __name__ == "__main__":
    sys.exit(run_sod_canary_test())

